File: examples/policies/python/list_and_filter.py
#!/usr/bin/env python3
"""
AxonFlow Policy Management - List and Filter Policies

VALIDATION: This example exits with code 1 if any assertion fails.

This example demonstrates how to:
- List all static policies
- Filter policies by category, tier, and status
- Get effective policies with tier inheritance

Run with: python list_and_filter.py
Prerequisites: docker compose up -d
"""

import asyncio
import os
import sys
from collections import Counter

from axonflow import (
    AxonFlow,
    ListStaticPoliciesOptions,
    PolicyCategory,
    PolicyTier,
)

failures: list[str] = []


def assert_check(condition: bool, message: str) -> None:
    """Check a condition and record failure if false."""
    if condition:
        print(f"   ✓ PASS: {message}")
    else:
        print(f"   ❌ FAIL: {message}")
        failures.append(message)


async def main() -> int:
    """List and filter policies."""
    print("AxonFlow Policy Management - List and Filter")
    print("=" * 60)

    client = AxonFlow(
        endpoint=os.getenv("AXONFLOW_ENDPOINT", "http://localhost:8080"),
        client_id="test-org-001",
        client_secret="test-secret",
    )

    try:
        # 1. List all policies
        print("\n1. Listing all policies...")

        all_policies = await client.list_static_policies()
        assert_check(all_policies is not None, "list_static_policies returned result")
        assert_check(isinstance(all_policies, list), "Result is a list")

        print(f"   Total: {len(all_policies)} policies")

        # Group by category for summary
        if all_policies:
            by_category = Counter(str(p.category.value) for p in all_policies)
            print("\n   By category:")
            for cat, count in by_category.items():
                print(f"     {cat}: {count}")

        # 2. Filter by category - SQL Injection policies
        print("\n2. Filtering by category (security-sqli)...")

        sqli_policies = await client.list_static_policies(
            ListStaticPoliciesOptions(category=PolicyCategory.SECURITY_SQLI)
        )
        assert_check(sqli_policies is not None, "Category filter returned result")
        assert_check(isinstance(sqli_policies, list), "Category filter result is a list")

        print(f"   Found: {len(sqli_policies)} SQLi policies")

        for p in sqli_policies[:3]:
            print(f"     - {p.name} (severity: {p.severity})")
        if len(sqli_policies) > 3:
            print(f"     ... and {len(sqli_policies) - 3} more")

        # 3. Filter by tier - System policies
        print("\n3. Filtering by tier (system)...")

        system_policies = await client.list_static_policies(
            ListStaticPoliciesOptions(tier=PolicyTier.SYSTEM)
        )
        assert_check(system_policies is not None, "Tier filter returned result")

        print(f"   Found: {len(system_policies)} system policies")

        # 4. Filter by enabled status
        print("\n4. Filtering by enabled status...")

        enabled_policies = await client.list_static_policies(
            ListStaticPoliciesOptions(enabled=True)
        )
        disabled_policies = await client.list_static_policies(
            ListStaticPoliciesOptions(enabled=False)
        )

        assert_check(enabled_policies is not None, "Enabled filter returned result")
        assert_check(disabled_policies is not None, "Disabled filter returned result")

        print(f"   Enabled: {len(enabled_policies)}")
        print(f"   Disabled: {len(disabled_policies)}")

        # 5. Combine filters
        print("\n5. Combining filters (enabled PII policies)...")

        pii_enabled = await client.list_static_policies(
            ListStaticPoliciesOptions(
                category=PolicyCategory.PII_GLOBAL,
                enabled=True,
            )
        )
        assert_check(pii_enabled is not None, "Combined filter returned result")

        print(f"   Found: {len(pii_enabled)} enabled PII policies")

        for p in pii_enabled[:5]:
            pattern_preview = p.pattern[:40] + "..." if len(p.pattern) > 40 else p.pattern
            print(f"     - {p.name}: {pattern_preview}")

        # 6. Get effective policies (includes tier inheritance)
        print("\n6. Getting effective policies...")

        effective = await client.get_effective_static_policies()
        assert_check(effective is not None, "get_effective_static_policies returned result")
        assert_check(isinstance(effective, list), "Effective policies is a list")

        print(f"   Effective total: {len(effective)} policies")

        if effective:
            by_tier = Counter(str(p.tier.value) for p in effective)
            print("\n   By tier (effective):")
            for tier, count in by_tier.items():
                print(f"     {tier}: {count}")

        # 7. Pagination example
        print("\n7. Pagination example...")

        page1 = await client.list_static_policies(
            ListStaticPoliciesOptions(limit=5, offset=0)
        )
        page2 = await client.list_static_policies(
            ListStaticPoliciesOptions(limit=5, offset=5)
        )

        assert_check(page1 is not None, "Pagination page 1 returned result")
        assert_check(page2 is not None, "Pagination page 2 returned result")
        assert_check(len(page1) <= 5, "Page 1 respects limit")
        assert_check(len(page2) <= 5, "Page 2 respects limit")

        print(f"   Page 1: {len(page1)} policies")
        print(f"   Page 2: {len(page2)} policies")

        # 8. Sorting
        print("\n8. Sorting by severity (descending)...")

        by_severity = await client.list_static_policies(
            ListStaticPoliciesOptions(
                sort_by="severity",
                sort_order="desc",
                limit=5,
            )
        )

        assert_check(by_severity is not None, "Sorting returned result")

        print("   Top 5 by severity:")
        for p in by_severity:
            print(f"     [{p.severity}] {p.name}")

    except Exception as e:
        print(f"\nError: {e}")
        failures.append(f"Policy listing failed: {e}")

    print("\n" + "=" * 60)
    if not failures:
        print("✓ ALL TESTS PASSED")
        print()
        print("Policy Listing validated:")
        print("  - list_static_policies() returns all policies")
        print("  - Category filtering works")
        print("  - Tier filtering works")
        print("  - Enabled/disabled filtering works")
        print("  - get_effective_static_policies() works")
        print("  - Pagination works")
        print("  - Sorting works")
        return 0
    else:
        print(f"❌ {len(failures)} TEST(S) FAILED:")
        for f in failures:
            print(f"   - {f}")
        return 1


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: examples/policies/typescript/list-and-filter.ts
/**
 * AxonFlow Policy Management - List and Filter Policies
 *
 * This example demonstrates how to:
 * - List all static policies
 * - Filter policies by category, tier, and status
 * - Get effective policies with tier inheritance
 *
 * VALIDATION: This example exits with code 1 if any assertion fails.
 */

import { AxonFlow } from '@axonflow/sdk';

const failures: string[] = [];

function assertCheck(condition: boolean, message: string): void {
  if (condition) {
    console.log(`   ✓ PASS: ${message}`);
  } else {
    console.log(`   ❌ FAIL: ${message}`);
    failures.push(message);
  }
}

async function main() {
  // Policy management APIs require clientId for X-Tenant-ID header
  const client = new AxonFlow({
    endpoint: process.env.AXONFLOW_ENDPOINT || 'http://localhost:8080',
    clientId: process.env.AXONFLOW_CLIENT_ID || 'demo-tenant',
  });

  console.log('AxonFlow Policy Management - List and Filter');
  console.log('='.repeat(60));

  try {
    // 1. List all policies
    console.log('\n1. Listing all policies...');

    const allPolicies = await client.listStaticPolicies();
    console.log(`   Total: ${allPolicies.length} policies`);

    assertCheck(Array.isArray(allPolicies), 'listStaticPolicies returns an array');
    assertCheck(allPolicies.length > 0, 'At least one policy exists in the system');

    // Group by category for summary
    const byCategory: Record<string, number> = {};
    allPolicies.forEach((p) => {
      byCategory[p.category] = (byCategory[p.category] || 0) + 1;
    });
    console.log('\n   By category:');
    Object.entries(byCategory).forEach(([cat, count]) => {
      console.log(`     ${cat}: ${count}`);
    });

    // 2. Filter by category - SQL Injection policies
    console.log('\n2. Filtering by category (security-sqli)...');

    const sqliPolicies = await client.listStaticPolicies({
      category: 'security-sqli',
    });
    console.log(`   Found: ${sqliPolicies.length} SQLi policies`);

    assertCheck(Array.isArray(sqliPolicies), 'Category filter returns an array');
    assertCheck(
      sqliPolicies.every(p => p.category === 'security-sqli'),
      'All filtered policies have security-sqli category'
    );

    // Show first 3
    sqliPolicies.slice(0, 3).forEach((p) => {
      console.log(`     - ${p.name} (severity: ${p.severity})`);
    });
    if (sqliPolicies.length > 3) {
      console.log(`     ... and ${sqliPolicies.length - 3} more`);
    }

    // 3. Filter by tier - System policies
    console.log('\n3. Filtering by tier (system)...');

    const systemPolicies = await client.listStaticPolicies({
      tier: 'system',
    });
    console.log(`   Found: ${systemPolicies.length} system policies`);

    assertCheck(Array.isArray(systemPolicies), 'Tier filter returns an array');
    assertCheck(
      systemPolicies.every(p => p.tier === 'system'),
      'All filtered policies have system tier'
    );

    // 4. Filter by enabled status
    console.log('\n4. Filtering by enabled status...');

    const enabledPolicies = await client.listStaticPolicies({
      enabled: true,
    });
    const disabledPolicies = await client.listStaticPolicies({
      enabled: false,
    });

    console.log(`   Enabled: ${enabledPolicies.length}`);
    console.log(`   Disabled: ${disabledPolicies.length}`);

    assertCheck(
      enabledPolicies.every(p => p.enabled === true),
      'All enabled-filtered policies have enabled=true'
    );
    assertCheck(
      disabledPolicies.every(p => p.enabled === false),
      'All disabled-filtered policies have enabled=false'
    );

    // 5. Combine filters
    console.log('\n5. Combining filters (enabled PII policies)...');

    const piiEnabled = await client.listStaticPolicies({
      category: 'pii-global',
      enabled: true,
    });
    console.log(`   Found: ${piiEnabled.length} enabled PII policies`);

    assertCheck(
      piiEnabled.every(p => p.category === 'pii-global' && p.enabled === true),
      'Combined filter returns policies matching both criteria'
    );

    piiEnabled.slice(0, 5).forEach((p) => {
      console.log(`     - ${p.name}: ${p.pattern.slice(0, 40)}...`);
    });

    // 6. Get effective policies (includes tier inheritance)
    console.log('\n6. Getting effective policies...');

    const effective = await client.getEffectiveStaticPolicies();
    console.log(`   Effective total: ${effective.length} policies`);

    assertCheck(Array.isArray(effective), 'getEffectiveStaticPolicies returns an array');
    assertCheck(effective.length > 0, 'At least one effective policy exists');

    // Group by tier
    const byTier: Record<string, number> = {};
    effective.forEach((p) => {
      byTier[p.tier] = (byTier[p.tier] || 0) + 1;
    });
    console.log('\n   By tier (effective):');
    Object.entries(byTier).forEach(([tier, count]) => {
      console.log(`     ${tier}: ${count}`);
    });

    // 7. Pagination example
    // Note: The backend uses page-based pagination (page=1, page=2, ...),
    // not offset-based. The SDK's offset param is not honored by the server,
    // so we verify that limit is respected and results are returned.
    console.log('\n7. Pagination example...');

    const page1 = await client.listStaticPolicies({
      limit: 5,
    });
    const page2 = await client.listStaticPolicies({
      limit: 3,
    });

    console.log(`   Page (limit=5): ${page1.length} policies`);
    console.log(`   Page (limit=3): ${page2.length} policies`);

    assertCheck(page1.length <= 5, 'Request with limit=5 returns at most 5 policies');
    assertCheck(page2.length <= 3, 'Request with limit=3 returns at most 3 policies');
    // When there are enough policies, verify that limit actually constrains results
    if (allPolicies.length > 5) {
      assertCheck(page1.length === 5, 'Limit of 5 returns exactly 5 when enough policies exist');
    }
    if (allPolicies.length > 3) {
      assertCheck(page2.length === 3, 'Limit of 3 returns exactly 3 when enough policies exist');
    }

    // 8. Sorting
    console.log('\n8. Sorting by severity (descending)...');

    const bySeverity = await client.listStaticPolicies({
      sortBy: 'severity',
      sortOrder: 'desc',
      limit: 5,
    });

    console.log('   Top 5 by severity:');
    bySeverity.forEach((p) => {
      console.log(`     [${p.severity}] ${p.name}`);
    });

    assertCheck(bySeverity.length <= 5, 'Sort query respects limit of 5');
    // Verify severity ordering (critical > high > medium > low)
    const severityOrder: Record<string, number> = { critical: 4, high: 3, medium: 2, low: 1 };
    let isOrdered = true;
    for (let i = 1; i < bySeverity.length; i++) {
      const prevSeverity = severityOrder[bySeverity[i - 1].severity] || 0;
      const currSeverity = severityOrder[bySeverity[i].severity] || 0;
      if (prevSeverity < currSeverity) {
        isOrdered = false;
        break;
      }
    }
    assertCheck(isOrdered, 'Policies are sorted by severity in descending order');

    console.log('\n' + '='.repeat(60));
    console.log('Example completed successfully!');

  } catch (error) {
    if (error instanceof Error) {
      console.error('\nError:', error.message);
    }
    failures.push(`Unexpected error: ${error instanceof Error ? error.message : error}`);
  }

  // Final assertion summary
  console.log('\n' + '='.repeat(60));
  console.log('Assertion Summary');
  console.log('='.repeat(60));
  if (failures.length === 0) {
    console.log('All assertions passed!');
  } else {
    console.log(`${failures.length} assertion(s) failed:`);
    failures.forEach((f) => console.log(`  - ${f}`));
  }

  process.exit(failures.length > 0 ? 1 : 0);
}

main();
